### ISSUE TITLE
Test Case Source us not available when using SetName

### ISSUE THREAD

**Issue by MattKeenum** [nunit/nunit3-vs-adapter#721]
NUnit: v3.12.0
NUnit3TestAdapter: v3.16.1
VS Edition: Professional 2019
VS Version: 16.4.4

This issue is probably related to Visual Studio rather than NUnit, but I thought you guys might have some insight.

**Issue: When using SetName and creating custom test cases, the source is no longer available in the Test Explorer.**

In a previous version of VS 2019, this was not the case. Test Cases created with SetName still had the original source (even when the Test Case was duplicated). This allowed the end user to double-click the custom Test Case and it would take them to the appropriate test method. In a recent update to VS 2019, the Test Explorer got an overhaul, and that is when this issue started happening.

The same issue happened In VS 2017. However, there was an extension installed and enabled by default: Dotnet Extension for Test Explorer. We could disable this extension and all of the custom test cases would then have a source attached to them. We could double-click a custom test case and it would take us to the test method. This extension is no longer showing in the Manage Extensions for VS 2019.

VS 2019
![1581087780825_image001](https://user-images.githubusercontent.com/8147082/74040116-b1777f80-4988-11ea-83d7-d1ec9eb2d892.png)

VS 2017
![1581087791493_image002](https://user-images.githubusercontent.com/8147082/74040131-b9372400-4988-11ea-9784-02b23fd6a9b5.png)

You can also see that in 2017 the traits were listed in alphabetical order, but not in 2019.

**Comment by OsirisTerje** [comment]
Yes, this is one of the issues we're struggling with now.  There is an alignment issue between the way the Test Explorer now works, and the adapter.  
For your case, you can try to use the two settings https://github.com/nunit/docs/wiki/Tips-And-Tricks#useparentfqnforparametrizedtests and https://github.com/nunit/docs/wiki/Tips-And-Tricks#usenunitidfortestcaseid and set both to true.  If may 1) work  or 2) crash or not run, depending upon your tests.  So YMMV. 

And yes, we're working on getting these things fixed.

**Comment by IbbuGit** [comment]
Hi,

I also faced the same issue and followed the setting that you have mentioned above. Added below lines in Run settings XML file.

  <NUnit>
    <UseParentFQNForParametrizedTests>true</UseParentFQNForParametrizedTests>
    <UseNUnitIdforTestCaseId>true</UseNUnitIdforTestCaseId>
  </NUnit>

I can get the source information but I am facing a different issue.

I am having multiple sets of data under one test method. For each data, I am naming different using SetName.

Now I am trying to run specific data from test explore but it runs all the data for the test method. Also, the spinner icon is not coming for the tests in the test explorer.

Kindly help me with this.

**Comment by OsirisTerje** [comment]
@IbbuGit When you turn on those settings, the effect is that the parent is being run, not the children.  So what you see is what is expected in that case.

**Comment by IbbuGit** [comment]
Oh ok. Is there any way I can run a specific child under the parent?

**Comment by OsirisTerje** [comment]
What happens if you dont turn those two settings on ?   With them off, you can run the testcases individually.   Thats the difference between those two.

**Comment by IbbuGit** [comment]
I am getting a source not available issue for each test data under the test method.

**Comment by IbbuGit** [comment]
Do we have any solution like I should not get source not available issues and I should be able to run the test case individually?
Currently I am using this   
    UseParentFQNForParametrizedTests-true
    UseNUnitIdforTestCaseId-true 
With this I am not facing source not available issue but not able to execute test individually.

**Comment by OsirisTerje** [comment]
1. This is part of the issue mentioned in 2nd comment above.  
2. Using the mentioned properties set to true, will make running by children of the method parent not possible. That's kind of the result of that fix, and also how Test Explorer expects this.

You should test without these properties, BUT turn off RTD (Real time discovery)(see Options/Test).  That way NUnit will handle the discovery and not Test Explorer, and it is more likely to hit correct.   I do believe this should fix your issue though.  
![image](https://user-images.githubusercontent.com/203432/81737694-cb00f980-9498-11ea-98f5-40dc284f1c17.png)

( Also, without knowing anything more about your specific source (no repro), I am just assuming how you have your code.  )

**Comment by IbbuGit** [comment]
I am getting below issue now

A source not available issue for each child(data) under the parent(test method) but able to execute a specific child under the parent.

**Comment by IbbuGit** [comment]
Any updates on this bug?

**Comment by OsirisTerje** [comment]
Sorry, still waiting for MS

**Comment by Kmaczek** [comment]
For me the same error happened when I did override ToString() for test data model. Some fields included in name had parenthesis ( ). Removing '(' and ')' fixed "No source available" problem.

Try to remove all the non letter characters, maybe its a valid workaround for this case too.
